src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,178 @@
-				ta[2] = xor_block(ad_offset, adp[2]);
-				ad_offset = xor_block(ad_offset, ctx->L[2]);
-				ta[3] = xor_block(ad_offset, adp[3]);
-				remaining -= 64;
-				k=4;
-			}
-			#endif
-			if (remaining >= 32) {
-				ad_offset = xor_block(ad_offset, ctx->L[0]);
-				ta[k] = xor_block(ad_offset, adp[k]);
-				ad_offset = xor_block(ad_offset, getL(ctx, ntz(k+2)));
-				ta[k+1] = xor_block(ad_offset, adp[k+1]);
-				remaining -= 32;
-				k+=2;
-			}
-			if (remaining >= 16) {
-				ad_offset = xor_block(ad_offset, ctx->L[0]);
-				ta[k] = xor_block(ad_offset, adp[k]);
-				remaining = remaining - 16;
-				++k;
-			}
-			if (remaining) {
-				ad_offset = xor_block(ad_offset,ctx->Lstar);
-				tmp.bl = zero_block();
-				memcpy(tmp.u8, adp+k, remaining);
-				tmp.u8[remaining] = (unsigned char)0x80u;
-				ta[k] = xor_block(ad_offset, tmp.bl);
-				++k;
-			}
-			AES_ecb_encrypt_blks(ta,k,&ctx->encrypt_key);
-			switch (k) {
-				#if (BPI == 8)
-				case 8: ad_checksum = xor_block(ad_checksum, ta[7]);
-					/* fallthrough */
-				case 7: ad_checksum = xor_block(ad_checksum, ta[6]);
-					/* fallthrough */
-				case 6: ad_checksum = xor_block(ad_checksum, ta[5]);
-					/* fallthrough */
-				case 5: ad_checksum = xor_block(ad_checksum, ta[4]);
-					/* fallthrough */
-				#endif
-				case 4: ad_checksum = xor_block(ad_checksum, ta[3]);
-					/* fallthrough */
-				case 3: ad_checksum = xor_block(ad_checksum, ta[2]);
-					/* fallthrough */
-				case 2: ad_checksum = xor_block(ad_checksum, ta[1]);
-					/* fallthrough */
-				case 1: ad_checksum = xor_block(ad_checksum, ta[0]);
-			}
-			ctx->ad_checksum = ad_checksum;
-		}
-	}
+int ae_encrypt(ae_ctx *ctx, const void *nonce_ptr, const void *pt_ptr,
+	       int pt_len, const void *ad_ptr, int ad_len, void *ct_ptr,
+	       void *tag, int final) {
+  assert(final == AE_FINALIZE);
+  int len = 0;
+  int ciphertext_len = 0;
+  const unsigned char *nonce = (const unsigned char *)nonce_ptr;
+  const unsigned char *pt = (const unsigned char *)pt_ptr;
+  const unsigned char* ad = (const unsigned char *)ad_ptr;
+  unsigned char* ct = (unsigned char *)ct_ptr;
+  // Set the key and nonce (IV) to the existing context if a nonce was
+  // provided. It will be omitted if `final` is set to AE_PENDING and
+  // "streaming" operation is in use.
+  if (nonce != NULL &&
+      EVP_EncryptInit_ex(ctx->enc_ctx, NULL, NULL, ctx->key, nonce) != 1) {
+    return -10;
+  }
+  if (ad != NULL && ad_len > 0 &&
+      EVP_EncryptUpdate(ctx->enc_ctx, NULL, &len, ad, ad_len) != 1) {
+     return -11;
+  }
+  len = 0;
+  if (pt != NULL && pt_len > 0 &&
+      EVP_EncryptUpdate(ctx->enc_ctx, ct, &len, pt, pt_len) != 1) {
+    return -12;
+  }
+  ciphertext_len += len;
+  if (EVP_EncryptFinal_ex(ctx->enc_ctx, ct + ciphertext_len, &len) != 1) {
+    return -13;
+  }
+  ciphertext_len += len;
+  // If `tag` is provided, the authentication tag goes
+  // there. Otherwise, it is appended after the ciphertext.
```

Comment:
Is putting the tag right after the ciphertext just a convention, or is it required by OCB mode? Probably worth commenting either way.